pkg/quay/quay_test.go

```diff
@@ -20,199 +20,461 @@ import (
 	robotName = "robot_name"
+
+	testRepoName        = "test-repo-using-api"
+	testRepoNamespace   = "redhat-appstudio-user"
+	testRepoDescription = "test repo"
 )
 
-var responseUnauthorized = []byte(`{"detail": "Unauthorized", "error_message": "Unauthorized", "error_type": "insufficient_scope", "title": "insufficient_scope", "type": "https://quay.io/api/v1/error/insufficient_scope", "status": 403}`)
+var responseUnauthorized = map[string]string{
+	"detail":        "Unauthorized",
+	"error_message": "Unauthorized",
+	"error_type":    "insufficient_scope",
+	"title":         "insufficient_scope",
+	"type":          "https://quay.io/api/v1/error/insufficient_scope",
+	"status":        "403",
+}
 
 func TestQuayClient_CreateRepository(t *testing.T) {
-	defer gock.Off()
+	testCases := []struct {
+		name               string
+		statusCode         int
+		responseData       interface{}
+		expectedRepository *Repository
+		expectedErr        string // Empty string means that no error is expected
+	}{
+		{
+			name:               "successful repository creation",
+			statusCode:         200,
+			responseData:       map[string]string{"name": testRepoName},
+			expectedRepository: &Repository{Name: testRepoName},
+			expectedErr:        "",
+		},
+		{
+			name:       "repository exists already",
+			statusCode: 400,
+			responseData: map[string]string{
+				"name":          testRepoName,
+				"error_message": "Repository already exists",
+			},
+			expectedRepository: &Repository{
+				Name:         testRepoName,
+				ErrorMessage: "Repository already exists",
+			},
+			expectedErr: "",
+		},
+		{
+			name:               "payment required",
+			statusCode:         402,
+			responseData:       map[string]string{"name": testRepoName},
+			expectedRepository: nil,
+			expectedErr:        "payment required",
+		},
+		{
+			name:       "not handled status with error message",
+			statusCode: 500, // can be any status code not handled by CreateRepository explicitly
+			responseData: map[string]string{
+				"name":          testRepoName,
+				"error_message": "something is wrong in the server",
+			},
+			expectedRepository: &Repository{
+				Name:         testRepoName,
+				ErrorMessage: "something is wrong in the server",
+			},
+			expectedErr: "something is wrong in the server",
+		},
+		{
+			name:               "response data can't be encoded to a JSON data",
+			statusCode:         200,
+			responseData:       "{\"name\": \"repo}",
+			expectedRepository: nil,
+			expectedErr:        "failed to unmarshal response",
+		},
+	}
 
-	gock.New("https://quay.io/api/v1").
-		MatchHeader("Content-type", "application/json").
-		MatchHeader("Authorization", "Bearer authtoken").
-		Post("/repository").
-		Reply(200).JSON(map[string]string{
-		"description": "description",
-		"namespace":   "redhat-appstudio-user",
-		"name":        "test-repo-using-api",
-	})
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			defer gock.Off()
```

Comment:
defer is called at the end of the function, so it can be just called once at beginning. Am I right @mmorhun

Reply:
Calling defer in the anonymous function intentionally due to some test function failed without it. But, I forget which one :sweat_smile: